dtm0/recovery.c

```diff
@@ -0,0 +1,2170 @@
+	 * lock will have a chance to continue.
+	 */
+	rc = m0_sm_timedwait(&m->rm_sm,
+			     M0_BITS(M0_DRMS_INIT, M0_DRMS_STOPPED),
+			     M0_TIME_NEVER);
+	M0_ASSERT_INFO(rc == 0, "rc=%d", rc);
+	recovery_machine_unlock(m);
+	unpopulate_foms(m);
+}
+
+static struct m0_reqh *
+m0_dtm0_recovery_machine_reqh(struct m0_dtm0_recovery_machine *m)
+{
+	return m->rm_svc->dos_generic.rs_reqh;
+}
+
+static const struct m0_fid *
+recovery_machine_local_id(const struct m0_dtm0_recovery_machine *m)
+{
+	return &m->rm_svc->dos_generic.rs_service_fid;
+}
+
+static int recovery_machine_log_iter_next(struct m0_dtm0_recovery_machine *m,
+					  struct m0_be_dtm0_log_iter *iter,
+					  const struct m0_fid *tgt_svc,
+					  const struct m0_fid *origin_svc,
+					  struct m0_dtm0_log_rec *record)
+{
+	M0_PRE(m->rm_ops.log_iter_next != NULL);
+	return m->rm_ops.log_iter_next(m, iter, tgt_svc, origin_svc, record);
+}
+
+static int recovery_machine_log_iter_init(struct m0_dtm0_recovery_machine *m,
+					  struct m0_be_dtm0_log_iter      *iter)
+{
+	M0_PRE(m->rm_ops.log_iter_init != NULL);
+	return m->rm_ops.log_iter_init(m, iter);
+
+}
+
+static void recovery_machine_log_iter_fini(struct m0_dtm0_recovery_machine *m,
+					   struct m0_be_dtm0_log_iter      *iter)
+{
+	M0_PRE(m->rm_ops.log_iter_fini != NULL);
+	m->rm_ops.log_iter_fini(m, iter);
+}
+
+static void recovery_machine_redo_post(struct m0_dtm0_recovery_machine *m,
+				       struct m0_fom                   *fom,
+				       const struct m0_fid *tgt_proc,
+				       const struct m0_fid *tgt_svc,
+				       struct dtm0_req_fop *redo,
+				       struct m0_be_op *op)
+{
+	M0_PRE(m->rm_ops.redo_post != NULL);
+	m->rm_ops.redo_post(m, fom, tgt_proc, tgt_svc, redo, op);
+}
+
+static void recovery_machine_recovered(struct m0_dtm0_recovery_machine *m,
+				       const struct m0_fid *tgt_proc,
+				       const struct m0_fid *tgt_svc)
+{
+	M0_PRE(m->rm_ops.ha_event_post != NULL);
+	m->rm_ops.ha_event_post(m, tgt_proc, tgt_svc,
+				M0_CONF_HA_PROCESS_DTM_RECOVERED);
+}
+
+static void recovery_machine_lock(struct m0_dtm0_recovery_machine *m)
+{
+	m0_sm_group_lock(&m->rm_sm_group);
+}
+
+static void recovery_machine_unlock(struct m0_dtm0_recovery_machine *m)
+{
+	m0_sm_group_unlock(&m->rm_sm_group);
+}
+
+enum recovery_fom_state {
+	RFS_INIT = M0_FOM_PHASE_INIT,
+	RFS_DONE = M0_FOM_PHASE_FINISH,
+	RFS_WAITING,
+	RFS_FAILED,
+	RFS_NR,
+};
+
+static struct m0_sm_state_descr recovery_fom_states[] = {
+	[RFS_INIT] = {
+		.sd_name      = "RFS_INIT",
+		.sd_allowed   = M0_BITS(RFS_WAITING, RFS_DONE, RFS_FAILED),
+		.sd_flags     = M0_SDF_INITIAL,
+	},
+	/* terminal states */
+	[RFS_DONE] = {
+		.sd_name      = "RFS_DONE",
+		.sd_allowed   = 0,
+		.sd_flags     = M0_SDF_TERMINAL,
+	},
+	/* failure states */
+	[RFS_FAILED] = {
+		.sd_name      = "RFS_FAILED",
+		.sd_allowed   = M0_BITS(RFS_DONE),
+		.sd_flags     = M0_SDF_FAILURE,
+	},
+
+	/* intermediate states */
+#define _ST(name, allowed)            \
```

Comment:
If this is only used for a single item, no need to define such macro.

I know similar macro is used in other places.

Reply:
3:Working on it.

Reply:
Fixed/merged.